File: Problems/0129. (M) Sum Root to Leaf Numbers/Java/solution.java
class Solution {
  public int sumNumbers(TreeNode root) {
    // recursive DFS

    return recurse(root, 0);
  }

  private int recurse(TreeNode root, int total) {
    if (root == null)
      return 0;

    total *= 10;
    total += root.val;
    int lr_sum = recurse(root.left, total) + recurse(root.right, total);

    return lr_sum > 0 ? lr_sum : total;
  }
}

File: Problems/0129. (M) Sum Root to Leaf Numbers/Python3/solution.py
class Solution:
  def sumNumbers(self, root: Optional[TreeNode], total=0) -> int:
    # recursive DFS

    if not root:
      return 0
    
    total *= 10
    total += root.val
    lr_sum = self.sumNumbers(root.left, total) + self.sumNumbers(root.right, total)

    return lr_sum if lr_sum else total

